### ISSUE TITLE
Cross-account ECR Support (Specify RegistryIds to GetAuthorizationTokenInput)

### ISSUE THREAD

**Issue by rmazlumyan** [kubernetes/kubernetes#28241]
I would like to be able to use AWS ECR registry across accounts.
Would it be feasible to add RegistryIds as input to GetAuthorizationTokenInput?

https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go#L194

**Comment by therc** [comment]
How would the account IDs make it to the credential provider?

**Comment by rmazlumyan** [comment]
account id is the host part of the registryURL right?

Similar to how region is being determined, couldn't the registryId be parsed from registryURLTemplate ?
const registryURLTemplate = "*.dkr.ecr.%s.amazonaws.com"

https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go#L51

**Comment by therc** [comment]
Even with the lazy credentials that I implemented, the ECR provider can't discover additional account IDs, because it does not know the full URL of the image. It just doesn't get passed around by the credential provider API. The match from image name to credentials happens outside of the ECR code. The lazy provider "only" helps us in not making each kubelet get, at startup, ECR credentials for all regions (most users will use at most one region, if any).

So, unless we change the whole credential API, the simplest solution might be to pass somehow a manual list of accounts to the kubelet. Ugly, I know.

**Comment by nagas** [comment]
+1 for ECS cross account support

**Comment by swladyka** [comment]
This seems to work now when policy per repo is set according to the docs: http://docs.aws.amazon.com/AmazonECR/latest/userguide/RepositoryPolicyExamples.html#IAM_allow_other_accounts

**Comment by Shemeikka** [comment]
At least for me this doesn't work using kubelet version:

```
Client Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.7", GitCommit:"095136c3078ccf887b9034b7ce598a0a1faff769", GitTreeState:"clean", BuildDate:"2017-07-05T16:51:56Z", GoVersion:"go1.7.6", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"6", GitVersion:"v1.6.2", GitCommit:"477efc3cbe6a7effca06bd1452fa356e2201e1ee", GitTreeState:"clean", BuildDate:"2017-04-19T20:22:08Z", GoVersion:"go1.7.5", Compiler:"gc", Platform:"linux/amd64"}
```


For this to work, RegistryIds needs to be added to the request and currently the credentialprovider doesn't do this.

[Todo: fill in RegistryIds?](https://github.com/kubernetes/kubernetes/blob/master/pkg/credentialprovider/aws/aws_credentials.go#L176)

**Comment by sambott** [comment]
I've started looking at this, at the moment I've added an optional list of additional ECR registryIds (AWS account numbers) to the cloud config. This way when it builds up the keychain it can add additional credential providers at this stage.

The nicer way, as @therc mentioned would be to change the DockerConfigProvider interface to use LazyProvide(url string) to allow the repo url to be passed to it. This way we could extract the registry id from the connection string before making the call to the AWS SDK's GetAuthorizationToken(input *ecr.GetAuthorizationTokenInput). I'm not sure how to gauge whether there would be appetite for a change as large as this when there is currently only one minor issue it would help with.

**Comment by macropin** [comment]
@sambott What is the current workaround?

**Comment by erickt** [comment]
@macropin: I haven’t tried it yet but you may be able to use https://github.com/upmc-enterprises/registry-creds to authenticate with a remote repository.

**Comment by sambott** [comment]
This was the current workaround. https://github.com/kubernetes/kubernetes/compare/release-1.8...sambott:cross_account_ecr

I haven't finished testing it at the moment, but feel free to make comments/suggestions. I'd be inclined to extend the LazyProvide that @therc implemented to extract registry ID from the URL instead though, I guess my worry is not knowing how big a rabbit hole that will end up being.

**Comment by macropin** [comment]
@erickt Thanks. I can confirm that https://github.com/upmc-enterprises/registry-creds is sufficient as a workaround until this issue is resolved and released.

**Comment by sambott** [comment]
Left it as the non-intrusive solution in the end. There is an additional, optional config variable that can override or introduce additional ECR registry IDs.

**Comment by sambott** [comment]
Not sure if anyone following this thread is able to code review the PR?

**Comment by kramarz** [comment]
Hey should it still be opened? I just copied my kubernetes setup to separate "development" aws account and It just worked. Only thing I had to do is to add policy for repositories in production account to allow development to pull images. It worked without any further configuration.
Kubernetes version v1.8.4

**Comment by sstarcher** [comment]
@kramarz same with v1.8.6 when adding a policy for a separate account Kubernetes cleanly worked.

**Comment by zkanda** [comment]
@kramarz @sstarcher hey guys, do you think give us how to do that or how it is actually structured?

For example, we have staging and prod account, the image we built is getting pushed to ECR in prod account, how do we configure staging to pull from prod?

**Comment by caioaao** [comment]
I see this was closed 1 year ago, but I can't find a solution online. Is this actually fixed or just rotten?

**Comment by TBBle** [comment]
Looks like it was fixed in March 2019 via #75587, which made the change described in https://github.com/kubernetes/kubernetes/issues/28241#issuecomment-341066009, so that the [DockerConfigProvider gets the URL it's looking up](https://github.com/kubernetes/kubernetes/pull/75587/files#diff-82e3d28f25a0fa876c9517c8e8814ca4), and can extract the registry ID.

It landed in k8s 1.15, but wasn't worth a changelog entry apparently.
